File: src/main/nodejs/Pipeline.ts
class Pipeline<T> {
    private data: T[] | T;

    constructor(data: T[] | T) {
        this.data = data;
    }

    static from<T>(data: T[] | T): Pipeline<T> {
        return new Pipeline(data);
    }

    where(predicate: (data: T) => boolean): Pipeline<T> {
        if (this.data instanceof Array) {
            const rs = this.data.filter(e => predicate(e));
            return new Pipeline<any>(rs);
        }

        if (predicate(this.data))
            return this
        else
            return new Pipeline<any>(null);
    }

    map<U>(fn: (data: T[] | T) => U): Pipeline<U> {
        if (this.data === null) {
            // If data is null, return a Pipeline with null as data
            return new Pipeline<U>(null as any);
        }
        const result = fn(this.data);
        return new Pipeline<U>(result);
    }

    first(): Pipeline<T> {
        if (this.data instanceof Array) {
            this.data = this.data[0];
        }
        return this;
    }

    get(): T[] | T {
        return this.data;
    }
}

File: src/main/java/org/app/common/pipeline/v0/Pipeline.java
package org.app.common.pipeline.v0;

import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;

public class Pipeline<T> {
    private final T data;

    public Pipeline(T data) {
        this.data = data;
    }

    public static <T> Pipeline<T> from(T source) {
        return new Pipeline<>(source);
    }

    public Pipeline<T> where(Predicate<T> predicate) {
        Optional<T> optional = Optional.of(data).filter(predicate);
        return optional.map(Pipeline::new)
                .orElseGet(() -> new Pipeline<>(null));
    }

    public <R> Pipeline<R> map(Function<T, R> transformer) {
        return Optional.of(data)
                .map(d -> new Pipeline<>(transformer.apply(d)))
                .orElseGet(() -> new Pipeline<>(null));
    }

    public Pipeline<T> then(Consumer<T> consumer) {
        consumer.accept(data);
        return this;
    }

    public static <T> Pipeline<T> sink(T source) {
        return new Pipeline<>(source);
    }

    public <R> Pipeline<R> sink(Function<T, R> function) {
        return new Pipeline<>(function.apply(data));
    }

    public T get() {
        return data;
    }
}